### PR TITLE
[cli] Implement upgrade compatibility checks client side

### PR DESCRIPTION
**Pull request by jordanjennings-mysten**
## Description 

Introduce client side upgrade compatibility checking, allowing users to check before TX submission if an upgrade will be successful. Improves the output of errors by relying on the move-binary-format crates checks to list the issues with the error to the user that are found.

## Test plan 

manually tested see comment below

---

## Release notes

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [x] CLI: 
User will see a different error when an upgrade error is thrown which includes the details of each error.
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

**Comment by jordanjennings-mysten** [comment]
# Tests
Still thinking about test strategy however I have run through some tests manually.
## UpgradeErrorsV1.move
```
/// Module: UpgradeErrors

#[allow(unused_field)]
module upgrades::upgrades {
    // struct missing
    public struct StructToBeRemoved {
        b: u64
    }

    // struct ability mismatch (add)
    public struct StructAbilityMismatchAdd {}

    // struct ability mismatch (remove)
    public struct StructAbilityMismatchRemove has copy {}

    // struct ability mismatch (change)
    public struct StructAbilityMismatchChange has copy {}

    // struct type param mismatch
    public struct StructTypeParamMismatch<S, T> { a: S }

    // struct field mismatch (add)
    public struct StructFieldMismatchAdd {
        a: u64,
        b: u64
    }

    // struct field mismatch (remove)
    public struct StructFieldMismatchRemove {
        a: u64,
        b: u64
    }

    // struct field mismatch (change)
    public struct StructFieldMismatchChange {
        a: u64,
        b: u64
    }

    // enum missing
    public enum EnumToBeRemoved {
        A,
        B
    }

    // enum ability mismatch (add)
    public enum EnumAbilityMismatchAdd  {
        A,
    }

    // enum ability mismatch (remove)
    public enum EnumAbilityMismatchRemove has copy {
        A,
    }

    // enum ability mismatch (change)
    public enum EnumAbilityMismatchChange has copy {
        A,
    }

    // enum new variant
    public enum EnumNewVariant {
        A,
        B,
        C
    }

    // enum variant missing
    public enum EnumVariantMissing {
        A,
        B,
    }

    // function missing public
    public fun function_to_have_public_removed() {}

    // function missing friend
    public(package) fun function_to_have_friend_removed() {}

    // function missing entry


    // function signature mismatch (add)
    public fun function_add_arg() {}

    // function signature mismatch (remove)
    public fun function_remove_arg(a: u64) {}

    // function signature mismatch (change)
    public fun function_change_arg(a: u64) {}
}
```

## Upgrade to -> UpgradeErrorsV2.move
```
/// Module: UpgradeErrors

#[allow(unused_field)]
module upgrades::upgrades {
    // struct missing
    // public struct StructToBeRemoved {}

    // struct ability mismatch (add)
    public struct StructAbilityMismatchAdd has copy {} // added the copy ability where none existed

    // struct field mismatch (remove)
    public struct StructAbilityMismatchRemove {} // removed the copy ability

    // struct field mismatch (change)
    public struct StructAbilityMismatchChange has drop {} // changed from drop to copy

    // struct type param mismatch
    public struct StructTypeParamMismatch<T> { a: T } // changed S to T

    // struct field mismatch (add)
    public struct StructFieldMismatchAdd {
        a: u64,
        b: u64,
        c: u64, // added
    }

    // struct field mismatch (remove)
    public struct StructFieldMismatchRemove {
        a: u64,
        // removed b: u64
    }

    // struct field mismatch (change)
    public struct StructFieldMismatchChange {
        a: u64,
        b: u8 // changed b from u64 to u8
    }

    // enum missing
    // public enum EnumToBeRemoved {}

    // enum ability mismatch (add)
    public enum EnumAbilityMismatchAdd has copy {
        A,
    }

    // enum ability mismatch (remove)
    public enum EnumAbilityMismatchRemove {
        A,
    }

    // enum ability mismatch (change)
    public enum EnumAbilityMismatchChange has drop {
        A,
    }

    // enum new variant
    public enum EnumNewVariant {
        A,
        B,
        C,
        D // new variant
    }

    // enum variant missing
    public enum EnumVariantMissing {
        A,
        // remove B,
    }

    // function missing public
    fun function_to_have_public_removed() {}

    // function missing friend
    fun function_to_have_friend_removed() {}

    // function missing entry

    // function signature mismatch (add)
    public fun function_add_arg(a: u64) {}

    // function signature mismatch (remove)
    public fun function_remove_arg() {}

    // function signature mismatch (change)
    public fun function_change_arg(a: u8) {} // now has u8 instead of u64
}
```

## Output
```
Upgrade compatibility check failed with the following errors:
- Struct ability mismatch: StructAbilityMismatchAdd
  Old: []
  New: [Copy, ]
- Struct ability mismatch: StructAbilityMismatchChange
  Old: [Copy, ]
  New: [Drop, ]
- Struct ability mismatch: StructAbilityMismatchRemove
  Old: [Copy, ]
  New: []
- Struct field mismatch: StructFieldMismatchAdd
  Old: [Field { name: Identifier("a"), type_: U64 }, Field { name: Identifier("b"), type_: U64 }]
  New: [Field { name: Identifier("a"), type_: U64 }, Field { name: Identifier("b"), type_: U64 }, Field { name: Identifier("c"), type_: U64 }]
- Struct field mismatch: StructFieldMismatchChange
  Old: [Field { name: Identifier("a"), type_: U64 }, Field { name: Identifier("b"), type_: U64 }]
  New: [Field { name: Identifier("a"), type_: U64 }, Field { name: Identifier("b"), type_: U8 }]
- Struct field mismatch: StructFieldMismatchRemove
  Old: [Field { name: Identifier("a"), type_: U64 }, Field { name: Identifier("b"), type_: U64 }]
  New: [Field { name: Identifier("a"), type_: U64 }]
- Struct missing: StructToBeRemoved
- Struct type param mismatch: StructTypeParamMismatch
  Old: [DatatypeTyParameter { constraints: [], is_phantom: false }, DatatypeTyParameter { constraints: [], is_phantom: false }]
  New: [DatatypeTyParameter { constraints: [], is_phantom: false }]
- Enum ability mismatch: EnumAbilityMismatchAdd
  Old: []
  New: [Copy, ]
- Enum ability mismatch: EnumAbilityMismatchChange
  Old: [Copy, ]
  New: [Drop, ]
- Enum ability mismatch: EnumAbilityMismatchRemove
  Old: [Copy, ]
  New: []
- Enum new variant: EnumNewVariant
  Old: [Variant { name: Identifier("A"), fields: [] }, Variant { name: Identifier("B"), fields: [] }, Variant { name: Identifier("C"), fields: [] }]
  New: [Variant { name: Identifier("A"), fields: [] }, Variant { name: Identifier("B"), fields: [] }, Variant { name: Identifier("C"), fields: [] }, Variant { name: Identifier("D"), fields: [] }]
- Enum missing: EnumToBeRemoved
  Enum { abilities: [], type_parameters: [], variants: [Variant { name: Identifier("A"), fields: [] }, Variant { name: Identifier("B"), fields: [] }] }
- Enum variant missing: EnumVariantMissing
  Variant { name: Identifier("B"), fields: [] }
- Function signature mismatch: function_add_arg
  Old:
    Params: []
    Return: []
  New:
    Params: [U64]
    Return: []
- Function signature mismatch: function_change_arg
  Old:
    Params: [U64]
    Return: []
  New:
    Params: [U8]
    Return: []
- Function signature mismatch: function_remove_arg
  Old:
    Params: [U64]
    Return: []
  New:
    Params: []
    Return: []
- Function lost friend visibility: function_to_have_friend_removed
- Function lost public visibility: function_to_have_public_removed
```

File: crates/sui/src/client_commands.rs
Comment (amnn) [inline crates/sui/src/client_commands.rs:871]: You should be able to convert the `chain_id` from the line about into a `Chain` enum.

File: crates/sui/src/upgrade_compatibility.rs
Comment (amnn) [inline crates/sui/src/upgrade_compatibility.rs:36]: `anyhow` has a `context` function that allows you to wrap an error in an `anyhow::Error`, adding extra contextual information -- it might be helpful to you here:

```suggestion
        .context("Unable to to deserialize compiled module")?;
```

File: crates/sui/src/upgrade_compatibility.rs
Comment (amnn) [inline crates/sui/src/upgrade_compatibility.rs:42]: ditto

```suggestion
        .context("Unable to get existing package")?;
```

File: crates/sui/src/upgrade_compatibility.rs
Comment (amnn) [inline crates/sui/src/upgrade_compatibility.rs:68]: nit: Parens should be unnecessary here

File: crates/sui/src/upgrade_compatibility.rs
Comment (amnn) [inline crates/sui/src/upgrade_compatibility.rs:66]: Might as well create a map from `new_modules` to avoid the scan each time.

File: crates/sui/src/upgrade_compatibility.rs
Comment (amnn) [reply crates/sui/src/upgrade_compatibility.rs:68]: Also, it would be good to make that change we talked about of having a constructor to build a `Compatibility` struct with all the right flags.

File: crates/sui/src/upgrade_compatibility.rs
Comment (amnn) [inline crates/sui/src/upgrade_compatibility.rs:1]: nit: It would be good to keep a standard file order:

- Types
- Impls
- Custom Trait Impls
- Stdlib Trait Impls
- Free functions

And in each section, `pub`, then `pub(crate)` then private.

Comment (amnn) [reply crates/sui/src/upgrade_compatibility.rs:1]: We don't adhere to this strictly, but worth trying to start as we mean to continue with new files 😅

File: crates/sui/src/upgrade_compatibility.rs
Comment (amnn) [inline crates/sui/src/upgrade_compatibility.rs:404]: You will need to pay attention to the `Compatibility` flags here, when deciding which errors to signal to people.

**Review by amnn — COMMENTED** [review]
The core of this change looks great -- one thing to note is that we do still need to pay attention to which checks are enabled in the `Compatibility` struct when we `finish`, but otherwise this logic looks good.

Main thing to work on before landing is to polish the content of error messages (and figure out the testing strategy).

File: crates/sui/src/client_commands.rs
Comment (jordanjennings-mysten) [reply crates/sui/src/client_commands.rs:871]: I think we might need this function. I'm having trouble finding it. searching the project for '35834a8a' mainnet id has some references but none lead me to a method of converting to the enum.

Comment (jordanjennings-mysten) [reply crates/sui/src/client_commands.rs:871]: I'm unsure about testnet since it gets wiped but maybe this?
https://github.com/jordanjennings-mysten/sui/commit/4cc83ef2438091768e2f5e2e982c88b57ff6d31d

File: crates/sui/src/upgrade_compatibility.rs
Comment (jordanjennings-mysten) [reply crates/sui/src/upgrade_compatibility.rs:1]: definitely makes things cleaner

File: crates/sui/src/client_commands.rs
Comment (amnn) [reply crates/sui/src/client_commands.rs:871]: This is what I was thinking of:

https://github.com/MystenLabs/sui/blob/54ccf09ef3764b5b94082860bcba858a15ce6ab7/crates/sui-types/src/digests.rs#L182

But it's a little annoying to use in this context, because you are given the chain identifier as a string, and that string only contains the first 4 bytes of the checkpoint digest. I think your solution (adding a new function for this) seems reasonable, but I would:

- Rely on the digests file to act as a source of truth for what the chain identifiers look like for mainnet and testnet so that if testnet is wiped, there's only a single place that needs to be updated.
- Add docs explaining why this function exists (really the issue is that the RPC Chain Identifier is contracted.

Comment (jordanjennings-mysten) [reply crates/sui/src/client_commands.rs]: I think this is resolved, shifted some things around. I think my change to digests makes the most sense given that `sui-protocol-config` is a dependency of `sui types`. In other words I cannot import the chain identifier b58 into the module where `Chain` exists because of a dependency loop so I pulled up the logic into sui types to match short hex ids.

File: crates/sui/src/upgrade_compatibility.rs
Comment (jordanjennings-mysten) [inline crates/sui/src/upgrade_compatibility.rs:476]: pass in OLD then NEW. fixes the issue we saw last week, falls under the tests.

File: crates/sui/Cargo.toml
Comment (amnn) [inline crates/sui/Cargo.toml:138]: To be honest, I would just pull the relevant changes out into their own commit so I could keep working on them locally, but land the rest of the stuff. Working on the formatting so we can enable this feature is going to be our immediate next step, so I wouldn't worry about checking in what you've already got (integration-wise) gated behind a feature.

File: crates/sui/src/unit_tests/fixtures/upgrade_errors/all_v1/Move.toml
Comment (amnn) [inline crates/sui/src/unit_tests/fixtures/upgrade_errors/all_v1/Move.toml:6]: Better to use a "local" dep for `Sui` in tests -- otherwise these tests will be slower and more flaky because they involve a network call out to fetch the git repo.

Comment (amnn) [reply crates/sui/src/unit_tests/fixtures/upgrade_errors/all_v1/Move.toml:6]: (ditto for all the other manifests)

File: crates/sui/src/unit_tests/upgrade_compatibility_tests.rs
Comment (amnn) [inline crates/sui/src/unit_tests/upgrade_compatibility_tests.rs:18]: I don't think the debug snapshot has quite worked here -- it's still converting to a string, and then it's showing the debug of that, which is not what we want. If we want `Debug` to work, I think you need to drop the `Error` impl for the error type and all the associated `#[error(...)]` annotations, (I think that's fine to do, because your proposed error format is not likely to re-use that) or alternatively, switch to a regular `assert_snapshot!` and use the current string format.

**Review by amnn — APPROVED** [review]
Thanks @jordanjennings-mysten -- comments are mostly clean-ups, but the most important one to address is the one about dependencies to `Sui` in test packages -- we should use a "local" dep for that, rather than a "git" dep to avoid making CI times longer.

Otherwise this is good to go, accepting to unblock!

File: crates/sui/Cargo.toml
Comment (jordanjennings-mysten) [reply crates/sui/Cargo.toml:138]: Yeah it just makes the git commands easier since I do like branching off of branches and continuing work. I have two branches off of compatibility-cli-mode (flags and now formatting) so it's preferable that I keep as much of the common in here.

The feature flag really is just to keep dead code warnings away.

Comment (jordanjennings-mysten) [reply crates/sui/Cargo.toml:138]: nevermind the flags liter a lot of the code, I'm going to just keep it out of the PR.

File: crates/sui/src/unit_tests/fixtures/upgrade_errors/all_v1/Move.toml
Comment (jordanjennings-mysten) [reply crates/sui/src/unit_tests/fixtures/upgrade_errors/all_v1/Move.toml]: turns out dont I need the dep

**Pull request merged**
